controllers/controllers.go

```diff
@@ -281,3 +296,55 @@ func markAsFreeOnSuccess(c client.Client, watchNamespace string, recorder record
 	}
 	return err
 }
+
+// markNodeAsUpgrading marks the given node as upgrading by adding an annotation to it. If the number of nodes
+// performing upgrades in parallel exceeds the maximum allowed, an error is returned.
+func markNodeAsUpgrading(ctx context.Context, c client.Client, node *core.Node) error {
+	controllerLocker.Lock()
+	defer controllerLocker.Unlock()
+
+	upgradingNodes, err := findUpgradingNodes(ctx, c, node)
+	if err != nil {
+		return err
+	}
+	if len(upgradingNodes) == 1 && upgradingNodes[0].Name == node.Name {
+		// node is already marked as upgrading
+		return nil
+	}
+
+	// check if marking the node as upgrading would exceed the maximum allowed
+	if len(upgradingNodes) >= MaxParallelUpgrades {
+		return fmt.Errorf("cannot mark node %s as upgrading, maximum number of parrallel upgrading nodes reached (%d)",
+			node.Name, MaxParallelUpgrades)
+	}
+	if err := metadata.ApplyUpgradingAnnotation(ctx, c, *node); err != nil {
+		return err
+	}
+	return nil
+}
+
+// findUpgradingNodes returns a list of nodes that are currently upgrading i.e. have the upgrading annotation set to
+// true. As shortcut-circuit if the given node is already marked as upgrading, it is returned as the only element
+// in the list.
+func findUpgradingNodes(ctx context.Context, c client.Client, node *core.Node) ([]core.Node, error) {
+	// get nodes with upgrading annotation
+	nodeList := &core.NodeList{}
+	if err := c.List(ctx, nodeList, client.MatchingLabels{core.LabelOSStable: "windows"}); err != nil {
+		return nil, fmt.Errorf("error listing Windows nodes: %w", err)
+	}
+	upgradingNodes := make([]core.Node, 0)
+	// check if any nodes are upgrading
+	for _, currentNode := range nodeList.Items {
+		v, ok := currentNode.Annotations[metadata.UpgradingAnnotation]
+		// check annotation present with value set to true
```

Comment:
unncessary comment, doesn't give any extra information that is not already obvious with `if ok && v == "true"`